Addresses bugs with gridcell averaging of FUN variables

wwieder opened the pull request:
Addresses issue #426 where special land units were assigned spval in InitCold for FUN variables.  As @billsacks noted in #431 "BGC variables are all set to 0 for special landunits via the SetValues subroutines in each *Type.F90 module". 

With missing values (instead of zeros), grid cell averaging did not work correctly, initially noticed over lakes where NFIX (patch level variable) > NFIX_TO_SMINN (the column level total).  PFT level output (h1) for NFIX matched grid cell totals (h0) for NFIX_TO_SMINN.

### Description of changes
Most changes in CNVegNitrogenFluxType.F90 including:
* Set these FUN variables to 0 for special land units. (This may be rudundant?)
* Reset spval to 0 in the Restart subroutine, to overwrite existing restart files.

Similar, but less extensive modifications need for CNVegCarbonFluxType.F90
Added conditional to SoilBiogeochemNitrogenFluxType.F90 where NFIX_TO_SMINN
is only on by default if FUN is off, as the variable is redundant with NFIX.

I also noticed there are other spvals defined in InitCold of CNVegCarbonFluxType.F90.  Here, I've only changed FUN related BGC variables, but are others handled consistently?

### Specific notes

Contributors other than yourself, if any:
@olyson, @ekluzek, @billsacks  

CTSM Issues Fixed (include github issue #):
#426 , addresses topics discussed in #431 

Are answers expected to change (and if so in what way)?
No science changes, but values from model diagnostics are different, as special land units will now receive zero values, not spval, and be weighted in grid cell averages accordingly- and consistently with other BGC variables.

Any User Interface Changes (namelist or namelist defaults changes)?
None

Testing performed, if any:
In an older tag, clm4_5_18_r270, I confirmed that special land units received 0 values with changes to InitCold. Separately, modifications to the Restart subroutine overwrote spval for NFIX, making results consistent with NFIX_TO_SMINN.  Tests with other FUN variables were not done, but are handled consistently.

Changes committed here did build correctly, but no simulations were done.

billsacks commented on src/biogeochem/CNVegCarbonFluxType.F90:
```diff
@@ -397,7 +397,6 @@ subroutine Init(this, bounds, carbon_type)
 
     call this%InitAllocate ( bounds, carbon_type)
     call this%InitHistory ( bounds, carbon_type )
-    call this%InitCold (bounds )
```
Why did you remove the call to InitCold here?

ekluzek replied:
@wweider, I think this is a mistake, since you have changes to InitCold. This essentially get's rid of InitCold. So I don't think this is right.

billsacks commented on src/biogeochem/CNVegCarbonFluxType.F90:
```diff
@@ -3574,6 +3571,7 @@ subroutine RestartBulkOnly ( this, bounds, ncid, flag )
             dim1name='pft', &
             long_name='', units='', &
             interpinic_flag='interp', readvar=readvar, data=this%leafc_to_litter_fun_patch)
+        call set_missing_vals_to_constant(this%leafc_to_litter_fun_patch, 0._r8)
```
For this and other new calls to set_missing_vals_to_constant: We (at least I) try to explicitly note where some code is only needed for the sake of backwards compatibility. See comments like this throughout the code:

```Fortran
! BACKWARDS_COMPATIBILITY(wjs, 2016-03-31)
```

This helps us understand what temporary code can possibly be removed a few years down the line. Can you please add these BACKWARDS_COMPATIBILITY notes (perhaps with a pointer to the issue and/or PR) to these new lines?

billsacks commented on src/soilbiogeochem/SoilBiogeochemNitrogenFluxType.F90:
```diff
@@ -316,10 +316,17 @@ subroutine InitHistory(this, bounds)
-         avgflag='A', long_name='symbiotic/asymbiotic N fixation to soil mineral N', &
-         ptr_col=this%nfix_to_sminn_col)
+    if ( use_fun ) then    ! redundant & confusing to write this out if FUN is on
+      this%nfix_to_sminn_col(begc:endc) = spval
```
It looks like this code is duplicated. It would be better to avoid this duplication, instead having something like:

```Fortran
if (use_fun) then
   default = 'inactive'
else
   default = 'active'
end if
call hist_addfld1d(..., default=default)

billsacks commented on src/soilbiogeochem/SoilBiogeochemNitrogenFluxType.F90:
```diff
@@ -316,23 +316,22 @@ subroutine InitHistory(this, bounds)
-           avgflag='A', long_name='symbiotic/asymbiotic N fixation to soil mineral N', &
-           ptr_col=this%nfix_to_sminn_col, default='inactive')
+    if (use_fun) then
+       default = 'inactive'
```
This variable needs to be declared

billsacks commented on src/soilbiogeochem/SoilBiogeochemNitrogenFluxType.F90:
```diff
@@ -316,23 +316,22 @@ subroutine InitHistory(this, bounds)
+    this%nfix_to_sminn_col(begc:endc) = spval
+    call hist_addfld1d (fname='NFIX_TO_SMINN', units='gN/m^2/s', &
+         avgflag='A', long_name='symbiotic/asymbiotic N fixation to soil mineral N', &
+         ptr_col=this%nfix_to_sminn_col)
```
Need to add `default=default` in this call

billsacks commented on src/soilbiogeochem/SoilBiogeochemNitrogenFluxType.F90:
```diff
@@ -316,23 +316,22 @@ subroutine InitHistory(this, bounds)
          avgflag='A', long_name='free living  N fixation to soil mineral N', &
          ptr_col=this%ffix_to_sminn_col)
 
+    default = 'active'
```
I don't think this is needed, since the `default` variable isn't used in the rest of this routine.

wwieder commented:
I think these issues have been corrected now?



On Thu, Jun 28, 2018 at 8:29 PM Bill Sacks <notifications@github.com> wrote:

> *@billsacks* requested changes on this pull request.
>
> Looks pretty good - thanks for addressing the comments. There are a few
> minor issues with the addition of the new default variable - see below.
> ------------------------------
>
> In src/soilbiogeochem/SoilBiogeochemNitrogenFluxType.F90
> <https://github.com/ESCOMP/ctsm/pull/432#discussion_r199038325>:
>
> > @@ -316,23 +316,22 @@ subroutine InitHistory(this, bounds)
>           avgflag='A', long_name='atmospheric N deposition to soil mineral N', &
>           ptr_col=this%ndep_to_sminn_col)
>
> -    if ( use_fun ) then    ! redundant & confusing to write this out if FUN is on
> -      this%nfix_to_sminn_col(begc:endc) = spval
> -      call hist_addfld1d (fname='NFIX_TO_SMINN', units='gN/m^2/s', &
> -           avgflag='A', long_name='symbiotic/asymbiotic N fixation to soil mineral N', &
> -           ptr_col=this%nfix_to_sminn_col, default='inactive')
> +    if (use_fun) then
> +       default = 'inactive'
>
> This variable needs to be declared
> ------------------------------
>
> In src/soilbiogeochem/SoilBiogeochemNitrogenFluxType.F90
> <https://github.com/ESCOMP/ctsm/pull/432#discussion_r199038390>:
>
> >      end if
> +    this%nfix_to_sminn_col(begc:endc) = spval
> +    call hist_addfld1d (fname='NFIX_TO_SMINN', units='gN/m^2/s', &
> +         avgflag='A', long_name='symbiotic/asymbiotic N fixation to soil mineral N', &
> +         ptr_col=this%nfix_to_sminn_col)
>
> Need to add default=default in this call
> ------------------------------
>
> In src/soilbiogeochem/SoilBiogeochemNitrogenFluxType.F90
> <https://github.com/ESCOMP/ctsm/pull/432#discussion_r199038428>:
>
> >
>      this%ffix_to_sminn_col(begc:endc) = spval
>      call hist_addfld1d (fname='FFIX_TO_SMINN', units='gN/m^2/s', &
>           avgflag='A', long_name='free living  N fixation to soil mineral N', &
>           ptr_col=this%ffix_to_sminn_col)
>
> +    default = 'active'
>
> I don't think this is needed, since the default variable isn't used in
> the rest of this routine.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/ctsm/pull/432#pullrequestreview-133080984>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHqLJHriTfa-9zBWoQG-uJ030qogH7SDks5uBZD5gaJpZM4U8FWV>
> .
>


-- 
Will Wieder
Project Scientist
CGD, NCAR
303-497-1352

ekluzek commented:
Since, this PR changes answers, it'll come in as it's own tag. We could include other answer changes with it as well. Right now the only other one is the use of the latest 1850 ndep file. So that will be added to the branch before it's brought in.

I also talked to Keith about if it really was desired to have non-veg land-units to be zero rather than missing values, and if averaging in those zero's was also desired. And the answer is yes, mainly because these are fluxes, so we want them defined everywhere and zeroed for any non veg landunits.

ekluzek commented:
Testing is complete and has gone as expected. Many tests change answers. Testing was NOT done on hobart, because of bug #433/

Here are the list of tests with differences and expected fails:

```
================================================================================
These tests compare different to the baseline
================================================================================
ERI_D_Ld9.f10_f10_musgs.I1850Clm50Bgc.cheyenne_gnu.clm-default
ERI_D_Ld9.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_gnu.clm-default
ERI_D_Ld9.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_intel.clm-default
ERI_D_Ld9.ne30_g16.I2000Clm50BgcCruGs.cheyenne_intel.clm-vrtlay
ERI_Ld9.f09_g17.I1850Clm50Bgc.cheyenne_intel.clm-drydepnomegan
ERI_Ld9.f09_g17.I2000Clm50BgcCruGs.cheyenne_intel.clm-cn_conly
ERI_Ld9.f09_g17.I2000Clm50BgcCruGs.cheyenne_intel.clm-drydepnomegan
ERI_Ld9.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_gnu.clm-default
ERI_Ld9.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_intel.clm-default
ERI_Ld9.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_intel.clm-drydepnomegan
ERI_Ld9.f45_g37.I2000Clm50BgcCruGs.cheyenne_intel.clm-nofire
ERI_N2_Ld9.f19_g17.I2000Clm50BgcCrop.cheyenne_intel.clm-default
ERP_D.f10_f10_musgs.IHistClm50Bgc.cheyenne_gnu.clm-decStart
ERP_D.f10_f10_musgs.IHistClm50Bgc.cheyenne_intel.clm-decStart
ERP_D_Ld3_P36x2.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_gnu.clm-default
ERP_D_Ld3_P36x2.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_intel.clm-default
ERP_D_Ld5.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_gnu.clm-ciso_flexCN_FUN
ERP_D_Ld5.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_intel.clm-rootlit
ERP_D_Ld5.f10_f10_musgs.IHistClm50BgcCrop.cheyenne_intel.clm-allActive
ERP_D_Ld5.f19_g17.I2000Clm50BgcCruGs.cheyenne_intel.clm-default
ERP_D_Ld5.f19_g17.IHistClm50Bgc.cheyenne_intel.clm-drydepnomegan
ERP_D_Ld5.f19_g17_gl4.I1850Clm50BgcCrop.cheyenne_intel.clm-glcMEC_changeFlags
ERP_D_Ld5.f19_g17_gl4.I1850Clm50BgcCropG.cheyenne_intel.clm-glcMEC_changeFlags
ERP_D_P36x2_Ld3.f10_f10_musgs.I1850Clm50BgcCrop.cheyenne_gnu.clm-default
ERP_D_P36x2_Ld3.f10_f10_musgs.I1850Clm50BgcCrop.cheyenne_intel.clm-default
ERP_D_P36x2_Ld3.f10_f10_musgs.I2000Clm50BgcCrop.cheyenne_intel.clm-cropColdStart
ERP_D_P36x2_Ld3.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_gnu.clm-default
ERP_D_P36x2_Ld3.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_intel.clm-default
ERP_D_P36x2_Ld3.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_intel.clm-flexCN_FUN
ERP_D_P36x2_Ld3.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_intel.clm-luna
ERP_D_P36x2_Ld30.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_intel.clm-default
ERP_D_P36x2_Ld5.f10_f10_musgs.I2000Clm50BgcCrop.cheyenne_intel.clm-irrig_spunup
ERP_D_P36x2_Ld5.f10_f10_musgs.I2000Clm50BgcCropRtm.cheyenne_intel.clm-irrig_spunup
ERP_Ld5.f09_g17.I1850Clm50BgcCropCru.cheyenne_intel.clm-ciso
ERP_Ld5.f10_f10_musgs.I1850Clm50Bgc.cheyenne_gnu.clm-default
ERP_Ld5.f10_f10_musgs.I1850Clm50Bgc.cheyenne_gnu.clm-drydepnomegan
ERP_Ld5.f10_f10_musgs.I1850Clm50Bgc.cheyenne_intel.clm-default
ERP_Ld5.f10_f10_musgs.I1850Clm50Bgc.cheyenne_intel.clm-drydepnomegan
ERP_Ld5.f19_g17.I1850Clm50Bgc.cheyenne_intel.clm-default
ERP_Ld5.f19_g17.I2000Clm50BgcCruGs.cheyenne_intel.clm-default
ERP_Ly3_P72x2.f10_f10_musgs.IHistClm50BgcCrop.cheyenne_intel.clm-cropMonthOutput
ERP_P180x2_D_Ld5.f19_g17.I2000Clm50BgcDvCrop.cheyenne_intel.clm-crop
ERP_P180x2_D_Ld5.f19_g17_gl4.I1850Clm50BgcCropG.cheyenne_intel.clm-glcMEC_increase
ERP_P36x2_D_Ld5.f10_f10_musgs.I1850Clm50Bgc.cheyenne_intel.clm-ciso
ERP_P36x2_Lm13.f10_f10_musgs.IHistClm50Bgc.cheyenne_gnu.clm-monthly
ERP_P36x2_Lm13.f10_f10_musgs.IHistClm50Bgc.cheyenne_intel.clm-monthly
ERP_P36x2_Lm25.f10_f10_musgs.I2000Clm50BgcDvCrop.cheyenne_intel.clm-monthly
ERP_P72x2_Lm25.f10_f10_musgs.I2000Clm50BgcDvCrop.cheyenne_intel.clm-monthly
ERP_P72x2_Lm36.f10_f10_musgs.I2000Clm50BgcCrop.cheyenne_intel.clm-clm50cropIrrigMonth_interp
ERS_D.f19_g17.I1850Clm50BgcCrop.cheyenne_intel.clm-reseedresetsnow
ERS_D_Ld12.f10_f10_musgs.I1850Clm50BgcCropG.cheyenne_intel.clm-glcMEC_spunup_inc_dec_bgc
ERS_D_Ld3.f09_g17_gl4.I1850Clm50BgcCrop.cheyenne_intel.clm-clm50KitchenSink
ERS_D_Ld3.f10_f10_musgs.I1850Clm50BgcCrop.cheyenne_gnu.clm-default
ERS_D_Ld3.f10_f10_musgs.I1850Clm50BgcCrop.cheyenne_intel.clm-default
ERS_D_Ld3.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_gnu.clm-default
ERS_D_Ld3.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_intel.clm-deepsoil_bedrock
ERS_D_Ld3.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_intel.clm-default
ERS_D_Ld3.f19_g17_gl4.I1850Clm50BgcCrop.cheyenne_intel.clm-clm50dynroots
ERS_D_Ld5.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_intel.clm-default
ERS_D_Ld7_Mmpi-serial.1x1_smallvilleIA.IHistClm50BgcCropGs.cheyenne_intel.clm-decStart1851_noinitial
ERS_Ld3.f09_g17.I1850Clm50BgcCrop.cheyenne_intel.clm-rad_hrly_light_res_half
ERS_Lm40_Mmpi-serial.1x1_numaIA.I2000Clm50BgcCropGs.cheyenne_gnu.clm-monthly
ERS_Lm40_Mmpi-serial.1x1_numaIA.I2000Clm50BgcCropGs.cheyenne_intel.clm-monthly
ERS_Lm54_Mmpi-serial.1x1_numaIA.I2000Clm50BgcCropGs.cheyenne_gnu.clm-cropMonthOutput
ERS_Lm54_Mmpi-serial.1x1_numaIA.I2000Clm50BgcCropGs.cheyenne_intel.clm-cropMonthOutput
ERS_Ly20_Mmpi-serial.1x1_numaIA.I2000Clm50BgcDvCropQianGs.cheyenne_gnu.clm-cropMonthOutput
ERS_Ly20_Mmpi-serial.1x1_numaIA.I2000Clm50BgcDvCropQianGs.cheyenne_intel.clm-cropMonthOutput
ERS_Ly3.f10_f10_musgs.I1850Clm50BgcCrop.cheyenne_intel.clm-clm50KSinkMOut
ERS_Ly3.f10_f10_musgs.I1850Clm50BgcCrop.cheyenne_intel.clm-cmip6
ERS_Ly3_Mmpi-serial.1x1_smallvilleIA.IHistClm50BgcCropGs.cheyenne_gnu.clm-cropMonthOutput
ERS_Ly3_Mmpi-serial.1x1_smallvilleIA.IHistClm50BgcCropGs.cheyenne_intel.clm-cropMonthOutput
ERS_Ly3_P72x2.f10_f10_musgs.IHistClm50BgcCropG.cheyenne_intel.clm-cropMonthOutput
ERS_Ly5_P144x1.f10_f10_musgs.IHistClm50BgcCrop.cheyenne_intel.clm-cropMonthOutput
ERS_Ly6_Mmpi-serial.1x1_smallvilleIA.IHistClm50BgcCropGs.cheyenne_gnu.clm-cropMonthOutput
ERS_Ly6_Mmpi-serial.1x1_smallvilleIA.IHistClm50BgcCropGs.cheyenne_intel.clm-cropMonthOutput
LII2FINIDATAREAS_D_P360x2_Ld1.f09_g16_gl4.I1850Clm50BgcCrop.cheyenne_intel.clm-compatible_finidat_f09
LII_D_Ld3.f19_g17_gl4.I2000Clm50BgcCrop.cheyenne_intel.clm-glcMEC_spunup_1way
LVG_Ld5_D.f10_f10_musgs.I1850Clm50Bgc.cheyenne_intel.clm-no_vector_output
PEM_D_Ld5.ne30_g16.I2000Clm50BgcCruGs.cheyenne_intel.clm-default
PET_P36x2_D.f10_f10_musgs.I1850Clm50BgcCrop.cheyenne_gnu.clm-default
PET_P36x2_D.f10_f10_musgs.I1850Clm50BgcCrop.cheyenne_intel.clm-default
SMS_D.f09_g16.I1850Clm50BgcSpinup.cheyenne_intel.clm-cplhist
SMS_D_Ld3.f10_f10_musgs.I1850Clm50BgcCrop.cheyenne_intel.clm-default
SMS_D_Ld3.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_gnu.clm-default
SMS_D_Ld3.f10_f10_musgs.I2000Clm50BgcCruGs.cheyenne_intel.clm-default
SMS_D_Ly2.1x1_brazil.IHistClm50BgcQianGs.cheyenne_intel.clm-ciso_bombspike1963
SMS_D_Ly2.1x1_numaIA.IHistClm50BgcCropGs.cheyenne_intel.clm-ciso_bombspike1963
SMS_Ld5.f19_g17.IHistClm50Bgc.cheyenne_intel.clm-decStart
SMS_Ld5_D.f09_g16.I1850Clm50BgcCrop.cheyenne_intel.clm-cmip6
SMS_Lm1.f09_g17_gl4.I1850Clm50Bgc.cheyenne_intel.clm-clm50KitchenSink
SMS_Lm1.f19_g17_gl4.I1850Clm50Bgc.cheyenne_intel.clm-clm50dynroots
SMS_Lm13.f19_g17.I2000Clm50BgcCrop.cheyenne_intel.clm-cropMonthOutput
SMS_Lm1_D.f10_f10_musgs.I2000Clm50BgcCrop.cheyenne_intel.clm-snowlayers_3_monthly
SMS_Ly3_Mmpi-serial.1x1_numaIA.I2000Clm50BgcCropGs.cheyenne_intel.clm-clm50dynroots
SMS_Ly3_Mmpi-serial.1x1_numaIA.I2000Clm50BgcDvCropQianGs.cheyenne_gnu.clm-cropMonthOutput
SSP_D_Ld10.f19_g17.I1850Clm50Bgc.cheyenne_intel.clm-rtmColdSSP
SSP_D_Ld4.f09_g17.I1850Clm50BgcCrop.cheyenne_intel.clm-ciso_rtmColdSSP
SSP_Ld10.f19_g17.I1850Clm50Bgc.cheyenne_intel.clm-rtmColdSSP
================================================================================
These tests failed
================================================================================
ERP_D_Ld10_P36x2.f10_f10_musgs.IHistClm50BgcCrop.cheyenne_intel.clm-ciso_decStart		EXPECTED
ERP_D_Ld5.f09_g17.I2000Clm50Vic.cheyenne_intel.clm-vrtlay		EXPECTED
ERS_Lm20_Mmpi-serial.1x1_smallvilleIA.I2000Clm50BgcCropGs.cheyenne_gnu.clm-monthly		EXPECTED
ERS_Lm20_Mmpi-serial.1x1_smallvilleIA.I2000Clm50BgcCropGs.cheyenne_intel.clm-monthly		EX
```
